Source/WebSocket_Windows.cpp

```diff
@@ -11,121 +11,98 @@ namespace UrlLib
-                    .then(arcana::inline_scheduler, arcana::cancellation::none(), [this]()
+                    .then(arcana::inline_scheduler, m_cancellationSource, [this]()
                     {
-                        m_readyState = ReadyState::Open;
-                        open_callback();
+                        m_onOpen();
+                    })
+                    .then(arcana::inline_scheduler, m_cancellationSource, [this](const arcana::expected<void, std::exception_ptr>& result)
+                    {
+                        if (result.has_error())
+                        {
+                            m_onError();
+                        }
                     });
             }
-            catch (hresult_error const& )
+            catch (hresult_error const&)
             {
-                error_callback();
+                m_onError();
             }
         }
 
         void Close()
         {
-            if (m_readyState == ReadyState::Closing ||
-                m_readyState == ReadyState::Closed)
-            {
-                error_callback();
-                throw std::runtime_error{"WebSocket is already Closing/Closed"};
-            }
-
-            m_readyState = ReadyState::Closing;
             m_webSocket.Close();
         }
 
     private:
         void OnWebSocketClosed(Windows::Networking::Sockets::IWebSocket const& /* sender */, Windows::Networking::Sockets::WebSocketClosedEventArgs const& args)
         {
-            m_readyState = ReadyState::Closed;
-            close_callback();
+            m_onClose();
         }
 
         void OnMessageReceived(Windows::Networking::Sockets::MessageWebSocket const& /* sender */, Windows::Networking::Sockets::MessageWebSocketMessageReceivedEventArgs const& args)
         {
-            if (m_readyState == ReadyState::Closed ||
-                m_readyState == ReadyState::Closing)
-            {
-                error_callback();
-                throw std::runtime_error{"WebSocket is Closing/Closed"};
-            }
-
             try
             {
                 Windows::Storage::Streams::DataReader dataReader{args.GetDataReader()};
                 dataReader.UnicodeEncoding(Windows::Storage::Streams::UnicodeEncoding::Utf8);
                 std::string message = winrt::to_string(dataReader.ReadString(dataReader.UnconsumedBufferLength()));
 
-                message_callback(message);
+                m_onMessage(message);
             }
             catch (winrt::hresult_error const& )
             {
-                error_callback();
+                m_onError();
             }
         }
        
         Windows::Networking::Sockets::MessageWebSocket m_webSocket;
-
-        std::function<void(void)> open_callback;
-        std::function<void(void)> close_callback;
-        std::function<void(std::string)> message_callback;
-        std::function<void(void)> error_callback;
+        arcana::cancellation_source m_cancellationSource{};
```

Comment:
It doesn't seem like there is code that cancels this source. You need to add code to do this in the destructor.